### PR TITLE
Add Travis CI for linux testing!

### PR DESCRIPTION
**Pull request by mbrandonw**
Thought this was gonna be simpler than it turned out, but I guess it makes sense that this would be kinda complicated. Will make some inline notes...

### CODE REVIEW

File: .swift-version [hunk 1]
```diff
@@ -0,0 +1 @@
+4.0
```
Comment (mbrandonw) [inline .swift-version]: using `swiftenv` on linux so need this...

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,9 @@
+FROM swift:4.0
+
+WORKDIR /package
+
+COPY . ./
+
+RUN swift package resolve
+RUN swift package clean
+CMD swift test --parallel
```
Comment (mbrandonw) [inline Dockerfile]: need this to run tests on linux locally...

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,2 @@
+test:
+	docker build --tag snapshot-testing . && docker run --rm snapshot-testing
```
Comment (mbrandonw) [inline Makefile]: now we can just `make test` locally to run on linux. maybe this should be `make test-linux`?

File: README.md [hunk 4]
```diff
@@ -1,4 +1,6 @@
-# swift-snapshot-testing [![CircleCI](https://circleci.com/gh/pointfreeco/swift-snapshot-testing.svg?style=svg)](https://circleci.com/gh/pointfreeco/swift-snapshot-testing)
+# swift-snapshot-testing
+
+Darwin [![CircleCI](https://circleci.com/gh/pointfreeco/swift-snapshot-testing.svg?style=svg)](https://circleci.com/gh/pointfreeco/swift-snapshot-testing) Linux [![Build Status](https://travis-ci.org/pointfreeco/swift-snapshot-testing.svg?branch=master)](https://travis-ci.org/pointfreeco/swift-snapshot-testing)
```
Comment (mbrandonw) [inline README.md]: this is what it looks like:

untitled

not happy with "Darwin"... not sure what to call it tho...

File: Sources/Diff/Diff.swift [hunk 5]
```diff
@@ -1,3 +1,5 @@
+import Foundation
```
Comment (mbrandonw) [inline Sources/Diff/Diff.swift]: need foundation to use `String.hasSuffx`

File: Sources/Diff/Diff.swift [hunk 6]
```diff
@@ -83,8 +85,8 @@ public struct Hunk {
   }
 }
 
-public func chunk<S: StringProtocol>(diff diffs: [Diff<S>], context ctx: Int = 4) -> [Hunk] {
```
Comment (mbrandonw) [inline Sources/Diff/Diff.swift]: `StringProtocol` doesnt have `hasSuffix` :/

File: Sources/SnapshotTesting/Diffable.swift [hunk 7]
```diff
@@ -2,6 +2,15 @@ import Diff
 import Foundation
 import XCTest
 
+#if os(Linux)
+  // NB: Linux doesn't have XCTAttachment, so stubbing out the minimal interface.
```
Comment (mbrandonw) [inline Sources/SnapshotTesting/Diffable.swift]: p much

File: Sources/SnapshotTesting/Diffable.swift [hunk 8]
```diff
@@ -39,8 +48,8 @@ extension String: Diffable {
     guard fst != snd else { return nil }
 
     let hunks = chunk(diff: diff(
-      fst.split(separator: "\n", omittingEmptySubsequences: false),
-      snd.split(separator: "\n", omittingEmptySubsequences: false)
+      fst.split(separator: "\n", omittingEmptySubsequences: false).map(String.init),
+      snd.split(separator: "\n", omittingEmptySubsequences: false).map(String.init)
```
Comment (mbrandonw) [inline Sources/SnapshotTesting/Diffable.swift]: doing this to get back into `String` world from `Substring` world

File: Sources/SnapshotTesting/SnapshotTesting.swift [hunk 9]
```diff
@@ -48,7 +48,7 @@ public func assertSnapshot<S: Snapshot>(
   }()
 
   let testName: String = {
-    let testIdentifier = "\(snapshotDirectoryUrl):\(function)"
+    let testIdentifier = "\(snapshotDirectoryUrl.absoluteString):\(function)"
```
Comment (mbrandonw) [inline Sources/SnapshotTesting/SnapshotTesting.swift]: linux's string convertible on `URL` prints the memory address.........

File: Tests/LinuxMain.swift [hunk 10]
```diff
@@ -0,0 +1,6 @@
+import XCTest
+@testable import SnapshotTestingTests
+
+XCTMain([
+    testCase(SnapshotTestingTests.allTests),
```
Comment (mbrandonw) [inline Tests/LinuxMain.swift]: gawd this stuff is gonna get out of sync so fast

File: Tests/SnapshotTestingTests/SnapshotTestingTests.swift [hunk 11]
```diff
@@ -39,3 +39,15 @@ class SnapshotTestingTests: XCTestCase {
...
   }
 }
+
+#if os(Linux)
+  extension SnapshotTestingTests {
+    static var allTests : [(String, (SnapshotTestingTests) -> () throws -> Void)] {
+      return [
+        ("testWithAny", testWithAny),
+        ("testWithNSObject", testWithNSObject),
+        ("testMultipleSnapshots", testMultipleSnapshots),
```
Comment (mbrandonw) [inline Tests/SnapshotTestingTests/SnapshotTestingTests.swift]: seriously?!

File: README.md (hunk 4 above)
Comment (stephencelis) [reply README.md]: macOS?

Comment (stephencelis) [reply README.md]: Ideally we run separately for macOS, iOS, tvOS, etc.

File: Sources/SnapshotTesting/Diffable.swift (hunk 8 above)
Comment (stephencelis) [reply Sources/SnapshotTesting/Diffable.swift]: :(

File: Sources/SnapshotTesting/SnapshotTesting.swift (hunk 9 above)
Comment (stephencelis) [reply Sources/SnapshotTesting/SnapshotTesting.swift]: :(

File: Sources/SnapshotTesting/SnapshotTesting.swift [hunk 12]
```diff
@@ -58,28 +58,33 @@ public func assertSnapshot<S: Snapshot>(
     .appendingPathExtension(pathExtension ?? "")
   let fileManager = FileManager.default
   try! fileManager.createDirectory(at: snapshotDirectoryUrl, withIntermediateDirectories: true)
-
+
```
Comment (stephencelis) [inline Sources/SnapshotTesting/SnapshotTesting.swift]: :(

File: README.md (hunk 4 above)
Comment (mbrandonw) [reply README.md]: true true! will do `macOS` for now but we should add tests for all platforms soon and then we can say macOS/iOS/tvOS...

**Comment by stephencelis** [comment]
Another possibility: move everything to Travis since it supports multiple envs. I wish Circle did :(

**Comment by mbrandonw** [comment]
> Looks good! Might be nice to call out each compromise made for Linux in a comment.

done!

> Another possibility: move everything to Travis since it supports multiple envs. I wish Circle did :(

im just worried about how unreliable travis can be, whereas circle seems to have gotten way more stable lately. plus circleci 2.0 will solve this for us, we just cant use it yet cause they dont support macos yet (I think 🤔)

**Comment by mbrandonw** [comment]
wooooo! our first Linus passes!

File: Tests/LinuxMain.swift (hunk 10 above)
Comment (kdawgwilk) [reply Tests/LinuxMain.swift]: Why not use sourcery to generate this?  Also you could add a unit test that only runs on macOS that uses the objective-c runtime to make sure every test is listed

Comment (mbrandonw) [reply Tests/LinuxMain.swift]: @kdawgwilk we for sure could do that, just ideally none of that would be necessary and will hopefully be improved soon!

**Comment by chefnobody** [comment]
> we just cant use it yet cause they dont support macos yet
FWIW, CCI's 2.0 product is available now on macOS containers: https://circleci.com/blog/one-more-thing-apple-developers-can-now-build-for-macos-ios-tvos-and-watchos-on-circleci-2-0/

This project looks awesome, btw.

**Comment by mbrandonw** [comment]
@chefnobody nice! we'll definitely look into just switching over to circle for everything...

**Pull request merged**